### ISSUE TITLE
get_realpath doesn't resolve symlinks in the last (basename) part of the input path

### ISSUE THREAD

**Issue by auno** [AsymLabs/realpath-lib#1]
Consider the following example:

``` bash
$ mkdir -p foo/bar1 foo/bar2
$ touch foo/bar1/baz
$ ln -s ../bar1/baz foo/bar2
$ ln -s bar1 foo/bar3
$ ls -l foo/bar*/baz
-rw-rw-r-- 1 auno auno  0 Nov  6 12:35 foo/bar1/baz
lrwxrwxrwx 1 auno auno 11 Nov  6 12:35 foo/bar2/baz -> ../bar1/baz
-rw-rw-r-- 1 auno auno  0 Nov  6 12:35 foo/bar3/baz

$ get_realpath foo/bar2/baz
/tmp/foo/bar2/baz
$ no_symlinks=true get_realpath foo/bar2/baz
/tmp/foo/bar2/baz
$ readlink -f foo/bar2/baz
/tmp/foo/bar1/baz

$ get_realpath foo/bar3/baz
/tmp/foo/bar3/baz
$ no_symlinks=true get_realpath foo/bar3/baz 
/tmp/foo/bar1/baz
```

get_realpath evidently only resolves symlinks in the dirname part of the given input path.

**Comment by AsymLabs** [comment]
This is an excellent test case and you are correct.  It should be possible to resolve it, I'll have a look at it.

The fault seems to be within `cd "$( echo "${1%/*}" )"` as it returns nothing for relative paths.

**Comment by AsymLabs** [comment]
As far as I can determine, there isn't a way to resolve a symlink to a file in pure Bash (as you already surmised).  It would appear that this is by _design_ (as to utilise posix/gnu tools for a given file system - although ksh offers some globbing possibilities).  The `get_realpath` function now incorporates `ls -dl` _only_ for the the special case of a linked file.  Note the following tests, yours plus some added complexity, ie a broken symlink (foo/bar2/broken) and an odd symlink (foo/bar1/foo->baz) where readlink leads to some strange results:  

```
[user@host ~]$ tree foo
foo
|-- bar1
|   |-- baz
|   `-- foo->baz
|-- bar2
|   |-- baz -> ../bar1/baz
|   |-- broken -> ../bar1/broken
|   `-- foobaz -> ../bar1/foo->baz
`-- bar3 -> bar1

3 directories, 5 files
```

And a helper function:  

```
[user@host ~]$ function status() { "$@"; code="$?"; (( $code )) && printf 'Failure : Error Code %s' $code; }
```

Now some test cases (with some odd results for readlink?):   

```
# Try existing and odd symlink directories and files.
[user@host ~]$ p='foo/bar3/baz'; unset no_symlinks; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")"
realpath: /home/user/foo/bar3/baz
readlink: /home/user/foo/bar1/baz
[user@host ~]$ p='foo/bar3/baz'; no_symlinks=true; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")"
realpath: /home/user/foo/bar1/baz
readlink: /home/user/foo/bar1/baz
[user@host ~]$ p='foo/bar2/foobaz'; unset no_symlinks; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")"
realpath: /home/user/foo/bar2/foobaz
readlink: /home/user/foo/bar1/foo->baz
[user@host ~]$ p='foo/bar2/foobaz'; no_symlinks=true; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")"
realpath: /home/user/foo/bar1/foo->baz
readlink: /home/user/foo/bar1/foo->baz

# Try resolving in local files.
[user@host ~]$ cd foo/bar2
[user@host bar2]$ p='baz'; unset no_symlinks; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")"
realpath: /home/user/foo/bar2/baz
readlink: /home/user/foo/bar1/baz
[user@host bar2]$ p='baz'; no_symlinks=true; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")"
realpath: /home/user/foo/bar1/baz
readlink: /home/user/foo/bar1/baz

# Try resolving in linked directory
[user@host bar2]$ cd ../bar3
[user@host bar3]$ p='baz'; unset no_symlinks; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")"
realpath: /home/user/foo/bar3/baz
readlink: /home/user/foo/bar1/baz
[user@host bar3]$ p='baz'; no_symlinks=true; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")" 
realpath: /home/user/foo/bar1/baz
readlink: /home/user/foo/bar1/baz

# Try path that does not exist: 'foo/bar2/nofoo'
[user@host bar3]$ cd ../../
[user@host ~]$ p='foo/bar2/nofoo'; unset no_symlinks; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")"
realpath: Failure : Error Code 1
readlink: /home/user/foo/bar2/nofoo # strange - this file does not exist!
[user@host ~]$  p='foo/bar2/nofoo'; no_symlinks=true; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")"
realpath: Failure : Error Code 1
readlink: /home/user/foo/bar2/nofoo # strange - this file does not exist!

# Try broken symlink: 'foo/bar2/broken'
[user@host ~]$  p='foo/bar2/broken'; unset no_symlinks; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")"
realpath: Failure : Error Code 1
readlink: /home/user/foo/bar1/broken # strange - this file does not exist!
[user@host ~]$ p='foo/bar2/broken'; no_symlinks=true; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")"
realpath: Failure : Error Code 1
readlink: /home/user/foo/bar1/broken # strange - this file does not exist!

# Try files that are not symlinked: 'foo/bar1/baz'
[user@host ~]$ p='foo/bar1/baz'; unset no_symlinks; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")"
realpath: /home/user/foo/bar1/baz
readlink: /home/user/foo/bar1/baz
[user@host ~]$ p='foo/bar1/baz'; no_symlinks=true; echo "realpath: $(status get_realpath "$p")"; echo "readlink: $(status readlink -f "$p")"
realpath: /home/user/foo/bar1/baz
readlink: /home/user/foo/bar1/baz
```

The new get_realpath function structure is:  

``` bash
# _file "path" : print the file of a path.
function _file(){
    printf '%s' "${1##*/}"
}

# _cd "path" : change the directory of a path.
function _cd(){
    cd "$(printf '%s' "${1%/*}")" 2>/dev/null
}

# get_realpath "path" : echo the realpath as logical or physical.
function get_realpath(){
    # 0 : Good.
    # 1 : Fail : no file/symlink broken/not regular file errors.
    # 2 : Fail : unspecified physical file path construction error.
    # 3 : Fail : unspecified logical or physical directory construction error. 
    [[ ! -f "$1" ]] && return 1 # Code 1.
    if [[ -z "$no_symlinks" ]]
    then
        local _pwd='pwd'
    else
        local _pwd='pwd -P'
        if [[ -L "$1" ]]
        then
            local _ls="$( ls -dl "$1" 2>/dev/null )" &&
            local _link="$( printf '%s' "${_ls#*"${1} -> "}" )" &&
            echo "$( _cd "$1"; _cd "$_link" && $_pwd )"/"$(_file "$_link")" && return 0 || return 2 # Code 2.
        fi
    fi
    echo "$( _cd "$1"; $_pwd )"/"$(_file "$1")" && return 0 || return 3 # Code 3.
}

```

I haven't tested this on OSX, BSD and other Unix or Linux variants, the major change is the use of `ls -dl`, that has been Posix standardized - so this script should be portable. Please note that this is offered under the license of the library, that will be updated to include it.

Thanks for you comments and criticisms that have been very insightful and constructive.

**Comment by auno** [comment]
I've compiled an automated test from all of these test cases (and added a new one). I've tested it on OS X, Solaris and FreeBSD and the results are the same on each of them, so the portability aspect of get_realpath is fine. However, the new test case I added, a symlink to a symlink in the basename part of the input path, fails. To resolve symlinks to symlinks, you would have to do it either recursively or using a loop. However, in order not to risk getting stuck in an infinite loop, you would also need to do symlink loop detection, and this is where it's really starting to get cumbersome.

##### Test output:

```
# Try existing and odd symlink directories and files
Testing existing symlink foo/bar3/baz (no_symlinks=):                           Passed
Testing existing symlink foo/bar3/baz (no_symlinks=true):                       Passed
Testing existing symlink foo/bar2/foobaz (no_symlinks=):                        Passed
Testing existing symlink foo/bar2/foobaz (no_symlinks=true):                    Passed
Testing existing symlink foo/bar2/bazbaz (no_symlinks=):                        Passed
Testing existing symlink foo/bar2/bazbaz (no_symlinks=true):                    Failed (expected "/tmp/foo/bar1/baz" but got "/baz")

# Try resolving in local files
Testing existing symlink baz (no_symlinks=):                                    Passed
Testing existing symlink baz (no_symlinks=true):                                Passed

# Try resolving in linked directory
Testing existing symlink baz (no_symlinks=):                                    Passed
Testing existing symlink baz (no_symlinks=true):                                Passed

# Try path that does not exist
Testing non-existant symlink foo/bar2/nofoo (no_symlinks=):                     Passed
Testing non-existant symlink foo/bar2/nofoo (no_symlinks=true):                 Passed

# Try broken symlink
Testing broken symlink foo/bar2/broken (no_symlinks=):                          Passed
Testing broken symlink foo/bar2/broken (no_symlinks=true):                      Passed

# Try files that are not symlinked
Testing ordinary file foo/bar1/baz (no_symlinks=):                              Passed
Testing ordinary file foo/bar1/baz (no_symlinks=true):                          Passed
```

##### get_realpath_test.sh:

``` bash
#!/bin/bash

. get_realpath.sh

_PWD="$(pwd)"
_PWDP="$(pwd -P)"

function status() { "$@"; code="$?"; (( $code )) && printf 'Error Code %s' $code; }

function test() {
  local message="$1"
  local path="$2"
  local expected="$3"

  if [[ "$expected" != "Error Code "* ]]; then
    if [[ -z "$no_symlinks" ]]; then
      expected="$_PWD/$expected"
    else
      expected="$_PWDP/$expected"
    fi
  fi

  printf "%-80s" "Testing $message $path (no_symlinks=$no_symlinks):"
  result="$(status get_realpath $path)"
  if [ "$result" = "$expected" ]; then
    echo "Passed"
  else
    echo "Failed (expected \"$expected\" but got \"$result\")"
  fi
}

echo "# Try existing and odd symlink directories and files"

                 test "existing symlink" "foo/bar3/baz" "foo/bar3/baz"
no_symlinks=true test "existing symlink" "foo/bar3/baz" "foo/bar1/baz"
                 test "existing symlink" "foo/bar2/foobaz" "foo/bar2/foobaz"
no_symlinks=true test "existing symlink" "foo/bar2/foobaz" "foo/bar1/foo->baz"
                 test "existing symlink" "foo/bar2/bazbaz" "foo/bar2/bazbaz"
no_symlinks=true test "existing symlink" "foo/bar2/bazbaz" "foo/bar1/baz"

echo
echo "# Try resolving in local files"

(
  cd foo/bar2
                   test "existing symlink" "baz" "foo/bar2/baz"
  no_symlinks=true test "existing symlink" "baz" "foo/bar1/baz"
)

echo
echo "# Try resolving in linked directory"

(
  cd foo/bar3
                   test "existing symlink" "baz" "foo/bar3/baz"
  no_symlinks=true test "existing symlink" "baz" "foo/bar1/baz"
)

echo
echo "# Try path that does not exist"

                 test "non-existant symlink" "foo/bar2/nofoo" "Error Code 1"
no_symlinks=true test "non-existant symlink" "foo/bar2/nofoo" "Error Code 1"

echo
echo "# Try broken symlink"

                 test "broken symlink" "foo/bar2/broken" "Error Code 1"
no_symlinks=true test "broken symlink" "foo/bar2/broken" "Error Code 1"

echo
echo "# Try files that are not symlinked"

                 test "ordinary file" "foo/bar1/baz" "foo/bar1/baz"
no_symlinks=true test "ordinary file" "foo/bar1/baz" "foo/bar1/baz"
```

##### setup_dirs.sh:

``` bash
#!/bin/bash

# Setting up the following structure:
# 
# foo
# ├── bar1
# │   ├── baz
# │   └── foo->baz
# ├── bar2
# │   ├── baz -> ../bar1/baz
# │   ├── bazbaz -> baz
# │   ├── broken -> ../bar1/broken
# │   └── foobaz -> ../bar1/foo->baz
# └── bar3 -> bar1

mkdir foo
mkdir foo/bar1
mkdir foo/bar2
ln -s bar1 foo/bar3
touch foo/bar1/baz
touch foo/bar1/foo-\>baz
ln -s ../bar1/baz foo/bar2/baz
ln -s baz foo/bar2/bazbaz
ln -s ../bar1/broken foo/bar2/broken
ln -s ../bar1/foo-\>baz foo/bar2/foobaz
```

**Comment by AsymLabs** [comment]
Thank you Auno - this is good progress - the issue of symlink loop detection needs more thought.  The **realpath-lib** repo is now substantially revised/updated - a more compact, modular approach that includes the improved `get_realpath` function and other improvements. The idea being that the `get_realpath` function forms the core and the other functions `get_dirname`, `get_filename` and `get_stemname` simply filter the results.

Your test script is very useful and should be added to the repo, if that is acceptable to you.

**Comment by auno** [comment]
Of course, consider it public domain.

**Comment by AsymLabs** [comment]
The script has been substantially revised and now environment settings are done to emulate the command **readlink -f** by default.  Two test scripts have been introduced 1) **make-generic-test.sh** and 2) **make-readlink-test.sh** (if readlink is installed on a system they are compared) that are based upon the concepts that you have illustrated above.  Tests now include the four functions **get_realpath**, **get_dirname**, **get_filename** and **get_stemname**.

The original test directory structure has been revised to include a circular symlink reference and to provide file name extensions.  The new directory structure is:

```
#   foo
#   ├── bar1
#   │   ├── baz.phys
#   │   ├── foo->bar1.sym -> ../bar2/foo->bar2.sym
#   │   └── foo->baz.phys
#   ├── bar2
#   │   ├── bazbaz.sym -> baz.sym 
#   │   ├── baz.sym -> ../bar1/baz.phys
#   │   ├── broken.sym -> ../bar1/broken.phys
#   │   ├── foo->bar2.sym -> ../bar1/foo->bar1.sym
#   │   └── foobaz.sym -> ../bar1/foo->baz.phys
#   └── bar3 -> bar1
```

Recursion is handed by brute force, in that each link that is traversed is stored within an array and this is unwound to stderror if either duplicates occur or when recursion limits are exceeded - so it could be used to diagnose deep symlink chain problems too.  Recursion has three exit conditions to prevent infinite loops, as the first encounter of either the **set_max_depth** environment setting, the built-in system recursion limit (40), or when duplicates are detected (circular reference).

The environment setting **no_symlinks** is now gone.  Instead the environments **set_logical**, **set_strict** and **set_max_depth** are provided.  These are explained in more detail in the README file.  By default the physical system is used so **set_logical** can be used to avoid this behaviour.   This is the opposite of the previous default environment.  Also strict file checking can be done by enabling **set_strict**.

It should be an easy task to test them on different systems and I have set them to produce test and error files for each given system that can be examined or posted if problems develop.

I cannot say that emulation to **readlink** is a certainty in all cases.  There may very well be some edge cases that are not yet tested or encountered.

**Comment by AsymLabs** [comment]
Just noted that the link history array was unrolled to stderr only when status 6 is thrown (circular reference) when it should be unrolled for status 4, 5 and 6 (recursion max environment, recursion depth limit built-in and circular reference, respectively).  So I've separated the operable code into a private function '_unroll' that takes the _paths array by reference and unrolls it to stderr.  This has now been incorporated into each of the three status conditions.

**Comment by AsymLabs** [comment]
Auno, thank you for your valuable input that has led to a much better outcome.  The thread has been dormant for some time so it should be closed now.

**Comment by auno** [comment]
Sorry for not getting back to you before. I lost track of this issue.

I've just run the tests on BSD, OS X and Solaris (as well as Linux of course). As standard BSD and OS X installs do not have GNU readlink installed (which is the reason for this lib to begin with), the readlink dependent tests failed on those platforms. I've updated the tests with hardcoded expected results though, after which they all pass.

I've created pull request #2 if you'd like to merge these updated tests.
